build/patches/Partitioning-Blink-memory-cache.patch

```diff
@@ -0,0 +1,194 @@
+ 
+   void Add(Resource*);
+   void Remove(Resource*);
+@@ -159,6 +157,8 @@ class PLATFORM_EXPORT MemoryCache final : public GarbageCollected<MemoryCache>,
+       base::MemoryPressureListener::MemoryPressureLevel) override;
+ 
+  private:
++   HeapVector<Member<Resource>> ResourcesForURL(const KURL&) const;
++
+   enum PruneStrategy {
+     // Automatically decide how much to prune.
+     kAutomaticPrune,
+diff --git a/third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc b/third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc
+--- a/third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc
++++ b/third_party/blink/renderer/platform/loader/fetch/resource_fetcher.cc
+@@ -640,7 +640,8 @@ Resource* ResourceFetcher::CreateResourceForStaticData(
+   if (!archive_ && factory.GetType() == ResourceType::kRaw)
+     return nullptr;
+ 
+-  const String cache_identifier = GetCacheIdentifier(url);
++  const String cache_identifier = GetCacheIdentifier(url,
++                          params.GetResourceRequest().TopFrameOrigin());
+   // Most off-main-thread resource fetches use Resource::kRaw and don't reach
+   // this point, but off-main-thread module fetches might.
+   if (IsMainThread()) {
+@@ -1083,7 +1084,9 @@ Resource* ResourceFetcher::RequestResource(FetchParameters& params,
+         resource = nullptr;
+       } else {
+         resource = GetMemoryCache()->ResourceForURL(
+-            params.Url(), GetCacheIdentifier(params.Url()));
++            params.Url(),
++            GetCacheIdentifier(params.Url(),
++              params.GetResourceRequest().TopFrameOrigin()));
+       }
+       if (resource) {
+         policy = DetermineRevalidationPolicy(resource_type, params, *resource,
+@@ -1291,7 +1294,8 @@ Resource* ResourceFetcher::CreateResourceForLoading(
+     const FetchParameters& params,
+     const ResourceFactory& factory) {
+   const String cache_identifier =
+-      GetCacheIdentifier(params.GetResourceRequest().Url());
++      GetCacheIdentifier(params.GetResourceRequest().Url(),
++        params.GetResourceRequest().TopFrameOrigin());
+   if (!base::FeatureList::IsEnabled(
+           blink::features::kScopeMemoryCachePerContext)) {
+     DCHECK(!IsMainThread() || params.IsStaleRevalidation() ||
+@@ -2185,13 +2189,16 @@ void ResourceFetcher::UpdateAllImageResourcePriorities() {
+   to_be_removed.clear();
+ }
+ 
+-String ResourceFetcher::GetCacheIdentifier(const KURL& url) const {
++String ResourceFetcher::GetCacheIdentifier(const KURL& url,
++              scoped_refptr<const blink::SecurityOrigin> origin) const {
++  String origin_url = origin ? origin->ToRawString() : "";
++
+   if (properties_->GetControllerServiceWorkerMode() !=
+       mojom::ControllerServiceWorkerMode::kNoController) {
+-    return String::Number(properties_->ServiceWorkerId());
++    return origin_url + "_" + String::Number(properties_->ServiceWorkerId());
```

Comment:
So if there is an URL:
```
https://www.example.com/mypage
```

and another one
```
https://www.example.com/mypage_1234
```

Then the first one with service worker id=1234 will collide?

Reply:
> Then the first one with service worker id=1234 will collide?

no, that change acts in `ResourceFetcher::GetCacheIdentifier` and not on the url, there are no changes to the url

Reply:
> will collide

but you are talking about the `CacheIdentifier` ...
that is, are you asking me if it is possible to bypass the identifier to reuse the cache?

Reply:
> If it is the same TLD+1 could it cause bugs related to wrong cache being used?

no, at most there is cache-miss

> In any case: why not use a different character than _ (that cannot exist on URLs) so that the problem is avoided altogether?

sounds good, then let's use space

Reply:
Better to use `|`? In printed outputs spaces can be missed.

Reply:
spaces, however, are already used to build the Network Isolation Key, perhaps it is better to follow that

Reply:
fixed in https://github.com/bromite/bromite/pull/2173/commits/216f4ce5c0bdad54a6e26ae0ca0f2965b9bf35fa